JBTM-3994  Fix for long running action to set right end status for a failure to compensate with at least 2 participant

arjantijms opened the pull request:
(the test doesn't fail now, but just prints the outcome)

For the context of this test, see https://issues.redhat.com/projects/JBTM/issues/JBTM-3994

arjantijms committed:
Added test for a failure to compensate with at least 2 participants

(the test doesn't fail now, but just prints the outcome)

marcosgopen commented on test/basic/src/test/java/io/narayana/lra/arquillian/FailedLRAMultiplePartipantsIT.java:
```diff
@@ -0,0 +1,202 @@
+            log.info("lra not in failed list (should have been)");
+
+            // TODO: fail test after fix
+            // fail("lra not in failed list");
```
Thanks @arjantijms for your reproducer. I will have a look at this. I would have expected this case to be covered in the TCK too. cc @xstefank

marcosgopen replied:
Looking at the server log I can see:
`2025-06-17 15:54:19,412 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-5) [0;1mStarting LRA with ID http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8ec0a_-3dbcef9e_68517332_2e9
2025-06-17 15:54:19,413 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-5) [0;1mReceived base URL from client: http://127.0.0.1:8180/FailedLRAMultiplePartipantsIT/
2025-06-17 15:54:19,414 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-5) [0;1mCalling participant 2 at http://127.0.0.1:8180/FailedLRAMultiplePartipantsIT/participant2/start-work
2025-06-17 15:54:19,429 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant2] (default task-1) [0;1m[34mJoining LRA with ID http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8ec0a_-3dbcef9e_68517332_2e9
2025-06-17 15:54:19,432 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-5) [0;1mReceived response from participant 2: http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8ec0a_-3dbcef9e_68517332_2e9
2025-06-17 15:54:19,432 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-5) [0;1mReturning status 500 to purposely fail the LRA!
2025-06-17 15:54:19,440 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant2] (default task-1) [0;1m[34m@Compensate called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8ec0a_-3dbcef9e_68517332_2e9. Returning 500 to purposely fail the compensate!
2025-06-17 15:54:19,441 WARN  [io.narayana.lra] (default task-2) LRAParticipantRecord: participant http://127.0.0.1:8180/FailedLRAMultiplePartipantsIT/participant2/compensate?method=jakarta.ws.rs.PUT reported a failure to compensate (cause FailedToCompensate)
2025-06-17 15:54:19,444 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) [0;1m@Compensate called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8ec0a_-3dbcef9e_68517332_2e9. Returning OK.
2025-06-17 15:54:19,453 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant2] (default task-1) [0;1m[34m@AfterLRA called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8ec0a_-3dbcef9e_68517332_2e9 with LRAStatus Cancelled
2025-06-17 15:54:19,456 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) [0;1m@AfterLRA called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8ec0a_-3dbcef9e_68517332_2e9 with LRAStatus Cancelled
2025-06-17 15:54:19,460 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-5) [0;1mgetLraStatus() called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8ec0a_-3dbcef9e_68517332_2e9.
2025-06-17 15:54:19,460 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-5) [0;1mAfterLRA called. Returning 200 with Cancelled `

arjantijms commented on test/basic/src/test/java/io/narayana/lra/arquillian/resource/LRAMultipleParticipant2.java:
```diff
@@ -0,0 +1,101 @@
+    public Response after(@HeaderParam(LRA_HTTP_ENDED_CONTEXT_HEADER) URI lraId, LRAStatus lraStatus) {
+        log.infov("\033[0;1m\u001B[34m@AfterLRA called for {0} with LRAStatus {1}", lraId, lraStatus);
+
+        return Response.ok().build();
```
I just tested it expicitly, and removing the AfterLRA method puts the LRA into failed as expected. I kinda saw that in the code, as the existence of the AfterLRA method is what caused the status "08" from a participant record to be lost effectively.

Of course, things should still work with an AfterLRA method present.

marcosgopen replied:
The behaviour is different if the @AfterLRA is present, in fact if that is present the participant is added in the heuristicList. cc @mmusgrov maybe you can explain it better than me why it is implemented so.

xstefank reviewed (commented):
I don't see anything wrong with this PR so I think that it should be merged.

@marcosgopen this cannot be ported to the spec since coordinator is not specified. We would need to introduce a generic mechanism to query failed LRAs.

marcosgopen commented on test/basic/src/test/java/io/narayana/lra/arquillian/resource/LRAMultipleParticipant1Initiator.java:
```diff
@@ -0,0 +1,141 @@
+        log.infov("\033[0;1mStarting LRA with ID {0}", lraId);
+        log.infov("\033[0;1mReceived base URL from client: {0}", baseUrl);
+
+        WebTarget target = ClientBuilder.newClient()
```
Having a closer look at your test I think you are involving the second participant in a wrong way, in fact if you want to have a second participant you can use the joinLRA method so that it is enlisted as a compensator for that LRA: https://github.com/jbosstm/lra/blob/61e9ed35c7c66acad4af11092476000a210fcca9/client/src/main/java/io/narayana/lra/client/internal/NarayanaLRAClient.java#L376.
I'll create a new test for this issue to explain you better what I mean.

marcosgopen replied:
Sorry @arjantijms, my bad. Your reproducer is straightforward and correct, I'll resolve this conversation

mmusgrov commented:
@arjantijms I am still investigating the wrong status reported by the AfterLRA  method call and the failures in other tests when I apply your recommended fix to LongRunningAction#endStateCheck ... please "bear with me" while I do that.

arjantijms commented:
> @arjantijms I am still investigating the wrong status reported by the AfterLRA method call and the failures in other tests when I apply your recommended fix to LongRunningAction#endStateCheck ... please "bear with me" while I do that.

Looking forward to hearing the results!

mmusgrov commented:
@arjantijms The problem is that when there is an @After annotation the participant record is not checking the failure status when it does the 'atEnd()' processing. So the fix is to return FINISH_ERROR if it failed:

```
[mmusgrov: lra] ((HEAD detached at pr/49)) $ git diff coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java
diff --git a/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java b/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java
index bb42c38fde..28143cbbf1 100644
--- a/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java
+++ b/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java
@@ -501,7 +501,8 @@ public class LRAParticipantRecord extends AbstractRecord implements Comparable<A
         // check the participant first since it will have been removed from one of the lists
         if (!isFinished() || !lra.isFinished()) {
             if (afterURI != null) {
-                return TwoPhaseOutcome.HEURISTIC_HAZARD;
+                if (isFailed())
+                    return isFailed() ? TwoPhaseOutcome.FINISH_ERROR : TwoPhaseOutcome.HEURISTIC_HAZARD;
             }
```

arjantijms committed:
Fix for long running action to set right end status

Set test to failing now if status would be wrong.

arjantijms commented:
> return isFailed() ? TwoPhaseOutcome.FINISH_ERROR : TwoPhaseOutcome.HEURISTIC_HAZARD;

Maybe that is the better fix. I only saw this after I just did the last commit (it actually lets the new test fail now, and did the alternative fix)

arjantijms commented:
> > > return isFailed() ? TwoPhaseOutcome.FINISH_ERROR : TwoPhaseOutcome.HEURISTIC_HAZARD;
> > 
> > 
> > Maybe that is the better fix. I only saw this after I just did the last commit (it actually lets the new test fail now, and did the alternative fix)
> 
> Basically returning FINISH_ERROR tells ArjunaCore to put the record on the failed list whereas returning HEURISTIC_HAZARD forces it back onto the heuristic list so we loose the FailedToCompensate status.

It sounded good, but it doesn't seem to work:

```
20:42:07,632 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Starting LRA with ID http://localhost:8080/lra-coordinator/lra-coordinator/0_ffff7f000001_d11d_685d947b_2
20:42:07,632 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Received base URL from client: http://127.0.0.1:8180/FailedLRAMultiplePartipantsIT/
20:42:07,633 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Calling participant 2 at http://127.0.0.1:8180/FailedLRAMultiplePartipantsIT/participant2/start-work
20:42:07,645 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant2] (default task-2) Joining LRA with ID http://localhost:8080/lra-coordinator/lra-coordinator/0_ffff7f000001_d11d_685d947b_2
20:42:07,654 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Received response from participant 2: http://localhost:8080/lra-coordinator/lra-coordinator/0_ffff7f000001_d11d_685d947b_2
20:42:07,654 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Returning status 500 to purposely fail the LRA!
20:42:07,727 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant2] (default task-2) @Compensate called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffff7f000001_d11d_685d947b_2. Returning 500 to purposely fail the compensate!
20:42:07,731 WARN  [io.narayana.lra] (default task-1) LRAParticipantRecord: participant http://127.0.0.1:8180/FailedLRAMultiplePartipantsIT/participant2/compensate?method=jakarta.ws.rs.PUT reported a failure to compensate (cause FailedToCompensate)
20:42:07,736 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-2) @Compensate called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffff7f000001_d11d_685d947b_2. Returning OK.
20:42:07,742 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-2) @AfterLRA called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffff7f000001_d11d_685d947b_2 with LRAStatus Cancelled
20:42:07,747 INFO  [io.narayana.lra] (default task-1) Failed lra record (Uid: 0:ffff7f000001:d11d:685d947b:2) moved to new location type: /StateManager/BasicAction/LongRunningAction/Failed
20:42:07,760 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) getLraStatus() called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffff7f000001_d11d_685d947b_2.
20:42:07,760 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) AfterLRA called. Returning 200 with Cancelled 
```

A working sequence would look like this:

```
21:16:39,502 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Starting LRA with ID http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_cd2a_685d9c93_2
21:16:39,503 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Received base URL from client: http://127.0.0.1:8180/FailedLRAMultiplePartipantsIT/
21:16:39,504 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Calling participant 2 at http://127.0.0.1:8180/FailedLRAMultiplePartipantsIT/participant2/start-work
21:16:39,521 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant2] (default task-2) Joining LRA with ID http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_cd2a_685d9c93_2
21:16:39,530 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Received response from participant 2: http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_cd2a_685d9c93_2
21:16:39,530 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Returning status 500 to purposely fail the LRA!
21:16:39,607 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant2] (default task-2) @Compensate called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_cd2a_685d9c93_2. Returning 500 to purposely fail the compensate!
21:16:39,611 WARN  [io.narayana.lra] (default task-1) LRAParticipantRecord: participant http://127.0.0.1:8180/FailedLRAMultiplePartipantsIT/participant2/compensate?method=jakarta.ws.rs.PUT reported a failure to compensate (cause FailedToCompensate)
21:16:39,615 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-2) @Compensate called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_cd2a_685d9c93_2. Returning OK.
21:16:39,623 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant2] (default task-2) @AfterLRA called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_cd2a_685d9c93_2 with LRAStatus FailedToCancel
21:16:39,627 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-2) @AfterLRA called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_cd2a_685d9c93_2 with LRAStatus FailedToCancel
21:16:39,629 INFO  [io.narayana.lra] (default task-1) Failed lra record (Uid: 0:ffffc0a8020e:cd2a:685d9c93:2) moved to new location type: /StateManager/BasicAction/LongRunningAction/Failed
21:16:39,642 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) getLraStatus() called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_cd2a_685d9c93_2.
21:16:39,642 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) AfterLRA called. Returning 200 with FailedToCancel 
```

Besides the obvious issue with the status being reported wrongly still, an additional problem with the other fix is that `@AfterLRA` of `LRAMultipleParticipant2` isn't called anymore. I guess this has to do with the conditional code for the `afterURI`. If present, it was set to heuristic, effectively making sure the participant is still being considered, and that seems to result in it ultimately being called again.

mmusgrov reviewed (approved):
@arjantijms Thanks for the test and for pointing out the flaw in my suggestion. Your fix does look good.  I'm approving the PR but let's wait for Marco to return from PTO before merging it.

mmusgrov commented on test/basic/src/test/java/io/narayana/lra/arquillian/FailedLRAMultiplePartipantsIT.java:
```diff
@@ -0,0 +1,203 @@
+
+        lrasToAfterFinish.add(lraId);
+
+        String status = getLraEndStatus(LRA_PARTICIPANT_PATH, LRA_END_STATUS, lraId);
```
@arjantijms Would it be possible to also check that LRAMultipleParticipant2 is also called correctly in the After method and with the correct status.

arjantijms replied:
I added the test for that. Good idea to test that!

The output is now:

```
23:03:52,041 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Starting LRA with ID http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_f6dd_68644d32_2
23:03:52,041 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Received base URL from client: http://127.0.0.1:8180/FailedLRAMultiplePartipantsIT/
23:03:52,043 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Calling participant 2 at http://127.0.0.1:8180/FailedLRAMultiplePartipantsIT/participant2/start-work
23:03:52,058 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant2] (default task-2) Joining LRA with ID http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_f6dd_68644d32_2
23:03:52,069 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Received response from participant 2: http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_f6dd_68644d32_2
23:03:52,069 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) Returning status 500 to purposely fail the LRA!
23:03:52,151 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant2] (default task-2) @Compensate called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_f6dd_68644d32_2. Returning 500 to purposely fail the compensate!
23:03:52,155 WARN  [io.narayana.lra] (default task-1) LRAParticipantRecord: participant http://127.0.0.1:8180/FailedLRAMultiplePartipantsIT/participant2/compensate?method=jakarta.ws.rs.PUT reported a failure to compensate (cause FailedToCompensate)
23:03:52,159 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-2) @Compensate called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_f6dd_68644d32_2. Returning OK.
23:03:52,167 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant2] (default task-2) @AfterLRA called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_f6dd_68644d32_2 with LRAStatus FailedToCancel
23:03:52,172 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-2) @AfterLRA called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_f6dd_68644d32_2 with LRAStatus FailedToCancel
23:03:52,174 INFO  [io.narayana.lra] (default task-1) Failed lra record (Uid: 0:ffffc0a8020e:f6dd:68644d32:2) moved to new location type: /StateManager/BasicAction/LongRunningAction/Failed
23:03:52,188 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) getLraStatus() called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_f6dd_68644d32_2.
23:03:52,189 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant1Initiator] (default task-1) AfterLRA called. Returning 200 with FailedToCancel 
INFO: [Jul 01, 2025 11:03:52] [io.narayana.lra.arquillian.FailedLRAMultiplePartipantsIT testWithStatusCompensateFailed] Received status 1 FailedToCancel (should be FailedToCancel)
23:03:52,192 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant2] (default task-1) getLraStatus() called for http://localhost:8080/lra-coordinator/lra-coordinator/0_ffffc0a8020e_f6dd_68644d32_2.
23:03:52,192 INFO  [io.narayana.lra.arquillian.resource.LRAMultipleParticipant2] (default task-1) AfterLRA called. Returning 200 with FailedToCancel 
INFO: [Jul 01, 2025 11:03:52] [io.narayana.lra.arquillian.FailedLRAMultiplePartipantsIT testWithStatusCompensateFailed] Received status 2 FailedToCancel (should be FailedToCancel)
```

marcosgopen commented:
Thanks very much for this PR @arjantijms , I am going to approve it but could I ask you to update the Jira and PR title to reflect that they are solving the bug and adding a test to verify that? e.g. "Fix for long running action to set right end status"

arjantijms commented:
Yes, I'll update the title and will look later today for the extra test. Thanks!

arjantijms committed:
Add testing of the end status of participant 2

arjantijms commented:
@mmusgrov @marcosgopen Hi, can this be merged now, or is there more to do still?
